File: Platform/infra/http/interfaces/ts/configuration.ts
/**
 * Configuration interfaces for HTTP client
 */

export enum RuntimeMode {
  PRODUCTION = 'production',
  REAL = 'real',
  TESTING = 'testing',
  MOCK = 'mock',
  DEBUG = 'debug',
  RECORD = 'record',
  REPLAY = 'replay',
  SIMULATION = 'simulation'
}

export interface HttpClientOptions {
  /** Default timeout in milliseconds */
  defaultTimeoutMs: number;
  /** User agent string */
  userAgent: string;
  /** Default headers to include in all requests */
  defaultHeaders: Record<string, string>;
  /** Maximum number of retries */
  maxRetries: number;
  /** Base URL for relative requests */
  baseUrl?: string;
}

export interface MockModeOptions {
  /** Default status code for mock responses */
  defaultStatusCode: number;
  /** Default response body */
  defaultBody: string;
  /** Default response headers */
  defaultHeaders: Record<string, string>;
  /** Simulate network latency in milliseconds */
  simulateLatencyMs: number;
}

export interface RecordModeOptions {
  /** Directory to store recorded responses */
  recordingPath: string;
  /** Whether to overwrite existing recordings */
  overwriteExisting: boolean;
  /** Include request headers in recordings */
  includeHeaders: boolean;
}

export interface ReplayModeOptions {
  /** Directory containing recorded responses */
  recordingPath: string;
  /** Behavior when no recording is found */
  fallbackMode: 'error' | 'passthrough' | 'mock';
  /** Whether to enforce strict URL matching */
  strictMatching: boolean;
}

export interface SimulationModeOptions {
  /** Path to simulation scenarios file */
  scenarioPath?: string;
  /** Global latency in milliseconds */
  globalLatencyMs: number;
  /** Global failure rate (0.0 to 1.0) */
  globalFailureRate: number;
  /** Minimum ping latency in milliseconds */
  minPingLatencyMs: number;
  /** Maximum ping latency in milliseconds */
  maxPingLatencyMs: number;
  /** Ping failure rate (0.0 to 1.0) */
  pingFailureRate: number;
}

export interface DebugModeOptions {
  /** Enable verbose logging */
  verboseLogging: boolean;
  /** Log request/response bodies */
  logBodies: boolean;
  /** Log headers */
  logHeaders: boolean;
}

export interface HttpClientModeOptions {
  /** Current runtime mode */
  mode: RuntimeMode;
  /** Mock mode configuration */
  mock: MockModeOptions;
  /** Record mode configuration */
  record: RecordModeOptions;
  /** Replay mode configuration */
  replay: ReplayModeOptions;
  /** Simulation mode configuration */
  simulation: SimulationModeOptions;
  /** Debug mode configuration */
  debug: DebugModeOptions;
}

export interface HttpClientConfig {
  /** HTTP client options */
  http: HttpClientOptions;
  /** Mode-specific options */
  mode: HttpClientModeOptions;
}

/**
 * Default configuration values
 */
export const DEFAULT_HTTP_OPTIONS: HttpClientOptions = {
  defaultTimeoutMs: 30000,
  userAgent: 'Coyote-HTTP-Client/1.0',
  defaultHeaders: {
    'Content-Type': 'application/json',
    'Accept': 'application/json'
  },
  maxRetries: 3
};

export const DEFAULT_MOCK_OPTIONS: MockModeOptions = {
  defaultStatusCode: 200,
  defaultBody: '{"message": "Mock response"}',
  defaultHeaders: {
    'Content-Type': 'application/json'
  },
  simulateLatencyMs: 100
};

export const DEFAULT_RECORD_OPTIONS: RecordModeOptions = {
  recordingPath: './recordings',
  overwriteExisting: false,
  includeHeaders: true
};

export const DEFAULT_REPLAY_OPTIONS: ReplayModeOptions = {
  recordingPath: './recordings',
  fallbackMode: 'error',
  strictMatching: true
};

export const DEFAULT_SIMULATION_OPTIONS: SimulationModeOptions = {
  globalLatencyMs: 100,
  globalFailureRate: 0.0,
  minPingLatencyMs: 10,
  maxPingLatencyMs: 100,
  pingFailureRate: 0.0
};

export const DEFAULT_DEBUG_OPTIONS: DebugModeOptions = {
  verboseLogging: true,
  logBodies: true,
  logHeaders: true
};

File: Platform/infra/http/src/dotnet/interfaces/Configuration.cs
namespace Coyote.Infra.Http;

/// <summary>
/// Runtime mode enumeration for the Coyote platform
/// </summary>
public enum RuntimeMode
{
    Production,
    Recording,
    Replay,
    Simulation,
    Debug,
    Testing
}

/// <summary>
/// Configuration options for HTTP client
/// </summary>
public class HttpClientOptions
{
    /// <summary>
    /// Default timeout in milliseconds
    /// </summary>
    public int DefaultTimeoutMs { get; set; } = 30000;
    
    /// <summary>
    /// Default headers to include in all requests
    /// </summary>
    public Dictionary<string, string> DefaultHeaders { get; set; } = new();
    
    /// <summary>
    /// Whether to verify server certificates
    /// </summary>
    public bool VerifyPeer { get; set; } = true;
    
    /// <summary>
    /// Whether to follow redirects
    /// </summary>
    public bool FollowRedirects { get; set; } = true;
    
    /// <summary>
    /// Client certificate path for mTLS
    /// </summary>
    public string? ClientCertPath { get; set; }
    
    /// <summary>
    /// Client certificate key path for mTLS
    /// </summary>
    public string? ClientKeyPath { get; set; }
    
    /// <summary>
    /// CA certificate path for server verification
    /// </summary>
    public string? CACertPath { get; set; }
    
    /// <summary>
    /// Base URL for relative requests
    /// </summary>
    public string? BaseUrl { get; set; }
    
    /// <summary>
    /// User agent string
    /// </summary>
    public string UserAgent { get; set; } = "Coyote-HttpClient/1.0";
    
    /// <summary>
    /// Maximum number of retries for failed requests
    /// </summary>
    public int MaxRetries { get; set; } = 3;
    
    /// <summary>
    /// Retry delay in milliseconds
    /// </summary>
    public int RetryDelayMs { get; set; } = 1000;
}

/// <summary>
/// Mode-specific configuration options
/// </summary>
public class HttpClientModeOptions
{
    /// <summary>
    /// Current runtime mode
    /// </summary>
    public RuntimeMode Mode { get; set; } = RuntimeMode.Production;
    
    /// <summary>
    /// Mock response configuration for testing mode
    /// </summary>
    public MockResponseOptions Mock { get; set; } = new();
    
    /// <summary>
    /// Debug mode configuration
    /// </summary>
    public DebugModeOptions Debug { get; set; } = new();
      /// <summary>
    /// Recording mode configuration
    /// </summary>
    public RecordingModeOptions Recording { get; set; } = new();
    
    /// <summary>
    /// Replay mode configuration
    /// </summary>
    public ReplayModeOptions Replay { get; set; } = new();
    
    /// <summary>
    /// Simulation mode configuration
    /// </summary>
    public SimulationModeOptions Simulation { get; set; } = new();
}

/// <summary>
/// Mock response configuration for testing
/// </summary>
public class MockResponseOptions
{
    /// <summary>
    /// Default mock response status code
    /// </summary>
    public int DefaultStatusCode { get; set; } = 200;
    
    /// <summary>
    /// Default mock response body
    /// </summary>
    public string DefaultBody { get; set; } = "{}";
    
    /// <summary>
    /// Default mock response headers
    /// </summary>
    public Dictionary<string, string> DefaultHeaders { get; set; } = new()
    {
        ["Content-Type"] = "application/json"
    };
    
    /// <summary>
    /// Simulated delay in milliseconds
    /// </summary>
    public int DelayMs { get; set; } = 0;
    
    /// <summary>
    /// Predefined responses for specific URLs
    /// </summary>
    public Dictionary<string, MockResponse> PredefinedResponses { get; set; } = new();
}

/// <summary>
/// Mock response definition
/// </summary>
public class MockResponse
{
    public int StatusCode { get; set; } = 200;
    public string Body { get; set; } = "{}";
    public Dictionary<string, string> Headers { get; set; } = new();
    public int DelayMs { get; set; } = 0;
}

/// <summary>
/// Debug mode configuration
/// </summary>
public class DebugModeOptions
{
    /// <summary>
    /// Whether to log request details
    /// </summary>
    public bool LogRequests { get; set; } = true;
    
    /// <summary>
    /// Whether to log response details
    /// </summary>
    public bool LogResponses { get; set; } = true;
    
    /// <summary>
    /// Whether to log headers
    /// </summary>
    public bool LogHeaders { get; set; } = true;
    
    /// <summary>
    /// Whether to log request/response bodies
    /// </summary>
    public bool LogBodies { get; set; } = true;
}

/// <summary>
/// Recording mode configuration
/// </summary>
public class RecordingModeOptions
{
    /// <summary>
    /// Path to store recorded requests/responses
    /// </summary>
    public string? RecordingPath { get; set; }
    
    /// <summary>
    /// Whether to record request bodies
    /// </summary>
    public bool RecordBodies { get; set; } = true;
    
    /// <summary>
    /// Whether to record headers
    /// </summary>
    public bool RecordHeaders { get; set; } = true;
}

/// <summary>
/// Replay mode configuration
/// </summary>
public class ReplayModeOptions
{
    /// <summary>
    /// Path to load recorded requests/responses from
    /// </summary>
    public string? RecordingPath { get; set; }
    
    /// <summary>
    /// Whether to use sequential replay mode (replay in order) or URL-based matching
    /// </summary>
    public bool SequentialMode { get; set; } = false;
    
    /// <summary>
    /// Whether to loop recordings when they're exhausted
    /// </summary>
    public bool LoopRecordings { get; set; } = false;
    
    /// <summary>
    /// What to do when no matching recording is found
    /// </summary>
    public ReplayFallbackMode FallbackMode { get; set; } = ReplayFallbackMode.DefaultResponse;
}

/// <summary>
/// Simulation mode configuration
/// </summary>
public class SimulationModeOptions
{
    /// <summary>
    /// Path to simulation scenario configuration file
    /// </summary>
    public string? ScenarioPath { get; set; }
    
    /// <summary>
    /// Global additional latency to add to all requests (ms)
    /// </summary>
    public int GlobalLatencyMs { get; set; } = 0;
    
    /// <summary>
    /// Global failure rate (0.0 to 1.0)
    /// </summary>
    public double GlobalFailureRate { get; set; } = 0.0;
    
    /// <summary>
    /// Minimum ping latency in milliseconds
    /// </summary>
    public int MinPingLatencyMs { get; set; } = 10;
    
    /// <summary>
    /// Maximum ping latency in milliseconds
    /// </summary>
    public int MaxPingLatencyMs { get; set; } = 100;
    
    /// <summary>
    /// Ping failure rate (0.0 to 1.0)
    /// </summary>
    public double PingFailureRate { get; set; } = 0.0;
}

/// <summary>
/// Replay fallback modes when no matching recording is found
/// </summary>
public enum ReplayFallbackMode
{
    /// <summary>
    /// Return a default 404 response
    /// </summary>
    DefaultResponse,
    
    /// <summary>
    /// Throw an exception
    /// </summary>
    ThrowException,
    
    /// <summary>
    /// Use the last available recording
    /// </summary>
    UseLastRecording
}
